src/pip/_internal/resolution/resolvelib/base.py

```diff
@@ -49,14 +51,19 @@ def __and__(self, other):
             return NotImplemented
         specifier = self.specifier & other.specifier
         hashes = self.hashes & other.hashes(trust_internet=False)
-        return Constraint(specifier, hashes)
+        links = self.links
+        if other.link:
+            links = links.union([other.link])
+        return Constraint(specifier, hashes, links)
 
     def is_satisfied_by(self, candidate):
         # type: (Candidate) -> bool
         # We can safely always allow prereleases here since PackageFinder
         # already implements the prerelease logic, and would have filtered out
         # prerelease candidates if the user does not expect them.
-        return self.specifier.contains(candidate.version, prereleases=True)
+        return self.specifier.contains(candidate.version, prereleases=True) and all(
+            _match_link(link, candidate) for link in self.links
+        )
```

Comment:
I would probably prefer doing this in a separate block instead, e.g.

```python
# Reject if there are any mismatched URL constraints on this package.
if self.links and all(_match_link(link, candidate) for link in self.links):
    return False
# We can safely always allow prereleases here since PackageFinder
# already implements the prerelease logic, and would have filtered out
# prerelease candidates if the user does not expect them.
return self.specifier.contains(candidate.version, prereleases=True)
```

It took me a couple of seconds that `all()` always returns `True` if `self.links` is empty, so the right side of `and` will pass if ther package has no URL constraints.

I don’t feel very strongly about this though, so feel free to not change this if you don’t want to. (Although an explanatory comment would be much appreciated.)